lib/semantics/symbol.cc

```diff
@@ -616,20 +618,53 @@ std::list<SourceName> DerivedTypeDetails::OrderParameterNames(
   return result;
 }
 
-std::list<Symbol *> DerivedTypeDetails::OrderParameterDeclarations(
+SymbolList DerivedTypeDetails::OrderParameterDeclarations(
     const Symbol &type) const {
-  std::list<Symbol *> result;
+  SymbolList result;
   if (const DerivedTypeSpec * spec{type.GetParentTypeSpec()}) {
     const DerivedTypeDetails &details{
         spec->typeSymbol().get<DerivedTypeDetails>()};
     result = details.OrderParameterDeclarations(spec->typeSymbol());
   }
-  for (Symbol *symbol : paramDecls_) {
+  for (const Symbol *symbol : paramDecls_) {
     result.push_back(symbol);
   }
   return result;
 }
 
+SymbolList DerivedTypeDetails::OrderComponents(const Scope &scope) const {
+  SymbolList result;
+  for (SourceName name : components_) {
+    auto iter{scope.find(name)};
+    if (iter != scope.cend()) {
+      const Symbol &symbol{*iter->second};
+      if (symbol.test(Symbol::Flag::ParentComp)) {
+        CHECK(result.empty());
+        const DerivedTypeSpec &spec{
+            *symbol.get<ObjectEntityDetails>().type()->AsDerived()};
```

Comment:
Since you're assuming the type is a derived type, you could use `symbol.get<ObjectEntityDetails>().type()->derivedTypeSpec()`. It asserts it is a derived type and returns a reference.

Maybe we shouldn't have both `AsDerived()` and `derivedTypeSpec()`.

Reply:
Thanks; fixed it here and in one other place that I found.